### PR TITLE
NNS1-3480: Fetch all transactions for an account

### PR DESCRIPTION
**Pull request by yhabib**
# Motivation

In the scope of feature NNS1-3480, we want to be able to fetch all transactions for a given `accountId`. The endpoint can't not return all transactions in one go so we have to keep calling until there are no more. 

It adds a configurable hard limit to the number of times it does the recursion.

# Changes

- Adds a new function `getAllTransactions` to fetch all transactions for a given `accountId`

# Tests

- Unit tests

# Todos

- [ ] Add entry to changelog (if necessary).
Not necessary

Next PR: #5873

### CODE REVIEW

File: frontend/src/lib/services/export-data.services.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  maxIterations?: number;
+  currentIteration?: number;
+}): Promise<TransactionWithId[] | undefined> => {
+  const identity = getCurrentIdentity();
+  if (identity instanceof AnonymousIdentity) return;
+
+  const maxResults = 100n;
+
+  try {
+    if (currentIteration > maxIterations) {
```
Comment (yhabib) [inline frontend/src/lib/services/export-data.services.ts:26]: I'll discuss with the PM how to treat this case or the error case but I handle it in a different PR.

File: frontend/src/tests/mocks/transaction.mock.ts [hunk 2]
```diff
@@ -45,6 +45,15 @@ export const mockTransactionWithId: TransactionWithId = {
   },
 };
 
+export const createMockTransactionWithId = (
```
Comment (yhabib) [inline frontend/src/tests/mocks/transaction.mock.ts:48]: I just saw there is a `createTransactionWithId` in `icp-transactions.mock.ts` so I'll delete this one :/

File: frontend/src/lib/services/export-data.services.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import { getTransactions } from "$lib/api/icp-index.api";
+import { type SignIdentity } from "@dfinity/agent";
+import type { TransactionWithId } from "@dfinity/ledger-icp";
+import { isNullish } from "@dfinity/utils";
+
+export const getAllTransactionsFromAccountAndIdentity = async ({
+  accountId,
+  identity,
+  start = undefined,
```
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:9]: Wouldn't it work the same w/o `= undefined`?

File: frontend/src/lib/services/export-data.services.ts (hunk 1 above)
Comment (mstrasinskis) [reply frontend/src/lib/services/export-data.services.ts:24]: Nitpick: Next time, you could add a TODO comment here.

File: frontend/src/lib/services/export-data.services.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  allTransactions = [],
+  currentIteration = 1,
+  maxIterations = 10,
```
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:12]: This could be a constant if there are no plans to have different values.

File: frontend/src/lib/services/export-data.services.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import { type SignIdentity } from "@dfinity/agent";
+import type { TransactionWithId } from "@dfinity/ledger-icp";
+import { isNullish } from "@dfinity/utils";
+
+export const getAllTransactionsFromAccountAndIdentity = async ({
+  accountId,
+  identity,
+  start = undefined,
+  allTransactions = [],
+  currentIteration = 1,
```
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:11]: Nitpick: How about using pageIndex or requestIndex instead of iteration to make it less generic?

File: frontend/src/lib/services/export-data.services.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        start: lastTx.id,
+        allTransactions: updatedTransactions,
+        maxIterations,
+        currentIteration: currentIteration + 1,
+      });
+    }
+
+    return updatedTransactions;
+  } catch (error) {
+    console.error("Error loading ICP account transactions:", error);
```
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:55]: Will this case be addressed in another PR? Alternatively, you could reuse the same message from [here](https://github.com/dfinity/nns-dapp/blob/1ead21f5e0b35b205a1e4b835399c834481514c7/frontend/src/lib/services/icp-transactions.services.ts#L39-L40).

File: frontend/src/lib/services/export-data.services.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
+import { getTransactions } from "$lib/api/icp-index.api";
+import { type SignIdentity } from "@dfinity/agent";
+import type { TransactionWithId } from "@dfinity/ledger-icp";
+import { isNullish } from "@dfinity/utils";
+
+export const getAllTransactionsFromAccountAndIdentity = async ({
```
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:6]: Personally, I would prefer exporting a cleaner version of this function (without the need for the recursion props) and another one to handle the dirty work. But it’s not a strong opinion.

File: frontend/src/lib/services/export-data.services.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    const { transactions, oldestTxId } = await getTransactions({
+      accountIdentifier: accountId,
+      identity,
+      maxResults,
+      start,
+    });
+
+    const updatedTransactions = [...allTransactions, ...transactions];
+    const completed =
+      isNullish(oldestTxId) || transactions.some(({ id }) => id === oldestTxId);
```
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:40]: If `isNullish(oldestTxId)` is needed only for the types, then `&&` might be a better fit here.

File: frontend/src/lib/services/export-data.services.ts (hunk 3 above)
Comment (mstrasinskis) [inline frontend/src/lib/services/export-data.services.ts:9]: Nitpick: how about `from` or `fromTransactionId`?

File: frontend/src/tests/lib/services/export-data.services.spec.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      });
+
+      const result = await getAllTransactionsFromAccountAndIdentity({
+        accountId: mockAccountId,
+        maxIterations: 3,
+        identity: mockSignInIdentity,
+      });
+
+      expect(result?.length).toBe(15);
+      expect(spyGetTransactions).toHaveBeenCalledTimes(3); // Should stop after second call
```
Comment (mstrasinskis) [inline frontend/src/tests/lib/services/export-data.services.spec.ts:103]: Or after third call? (the comment)

File: frontend/src/tests/lib/services/export-data.services.spec.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        })
+        .mockRejectedValueOnce(new Error("API Error"));
+
+      const result = await getAllTransactionsFromAccountAndIdentity({
+        accountId: mockAccountId,
+        identity: mockSignInIdentity,
+      });
+
+      expect(result).toEqual(firstBatch);
+      expect(spyGetTransactions).toHaveBeenCalledTimes(2);
```
Comment (mstrasinskis) [inline frontend/src/tests/lib/services/export-data.services.spec.ts:125]: We should definitely display something to the user, but I believe this will be addressed later.

**Review by mstrasinskis — APPROVED** [review]
Looks good, just a bunch of minor comments.

**Review by mstrasinskis — COMMENTED** [review]
Sorry, wrong selection. It would be nice to address a couple of comments.

File: frontend/src/lib/services/export-data.services.ts (hunk 8 above)
Comment (mstrasinskis) [reply frontend/src/lib/services/export-data.services.ts:40]: Or make it more obvious, like `isNullish(oldestTxId) ? true : transactions.some(({ id }) => id === oldestTxId);`. But feel free to ignore :)

File: frontend/src/lib/services/export-data.services.ts (hunk 7 above)
Comment (yhabib) [reply frontend/src/lib/services/export-data.services.ts:6]: The description should have been more clear. This function will not be exposed to consumers, I had to do it now for tests so I thought it would be ok like this but I'm open to create a wrapper around it.

There will be another function in this service that will take a list of addresses and invoke this function to get all NNS transactions for a given user.

File: frontend/src/lib/services/export-data.services.ts (hunk 3 above)
Comment (yhabib) [reply frontend/src/lib/services/export-data.services.ts:9]: yes, though I wanted to be very explicit.

Comment (yhabib) [reply frontend/src/lib/services/export-data.services.ts:9]: I followed the pattern of `getTransactions` but i can change it

File: frontend/src/lib/services/export-data.services.ts (hunk 5 above)
Comment (yhabib) [reply frontend/src/lib/services/export-data.services.ts:11]: Yes, I thought about using the concept of pages, but I didn't want to introduce the term to avoid confusion. However, I'm open to it.

File: frontend/src/lib/services/export-data.services.ts (hunk 4 above)
Comment (yhabib) [reply frontend/src/lib/services/export-data.services.ts]: Indeed.

File: frontend/src/tests/lib/services/export-data.services.spec.ts (hunk 10 above)
Comment (yhabib) [reply frontend/src/tests/lib/services/export-data.services.spec.ts:124]: Yes, there will be consumers for this function that will take care of that

File: frontend/src/lib/services/export-data.services.ts (hunk 8 above)
Comment (yhabib) [reply frontend/src/lib/services/export-data.services.ts:40]: I have this other one #5874 for the same. Lets agree with @dskloetd on what is the best and I'll ported it here.

**Review by mstrasinskis — APPROVED** [review]
Thank you!
